File: 1871-jump-game-vii/1871-jump-game-vii.cpp
class Solution {
public:
    bool canReach(string s, int minJump, int maxJump) {
        queue<int> q;
        q.push(0);
        
        int mais_longe = 0;
        int n = s.length();

        
        if (s.back() == '1') {
            return false;
        }
        if (maxJump == n - 1) {
            return true;
        }

        while (!q.empty()) {
            int i = q.front(); // Pega o primeiro elemento
            q.pop();           // Remove o primeiro elemento 

            
            int comeco = max(i + minJump, mais_longe + 1);
            
            // Calcula o limite do loop 
            int limite = min(i + maxJump + 1, n);

            for (int j = comeco; j < limite; ++j) {
                if (s[j] == '0') {
                    q.push(j);
                    if (j == n - 1) {
                        return true;
                    }
                }
            }
            
            mais_longe = i + maxJump;
        }

        return false;
    }
};

File: 1871-jump-game-vii/1871-jump-game-vii.py
class Solution(object):
    def canReach(self, s, minJump, maxJump):
        q = deque([0])
        mais_longe = 0
        if s[-1]=='1':
            return False
        if maxJump == len(s) - 1:
            return True
        while q:
            i = q.popleft()
            comeco = max(i + minJump, mais_longe + 1)

            for j in range (comeco, min(i + maxJump + 1, len(s))):
                if s[j] == "0":
                    q.append(j)
                    if j == len(s) - 1:
                        return True
            mais_longe = i + maxJump
        return False
